sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -648,8 +648,13 @@ class Analyzer(
           if catalog.isTemporaryTable(ident) =>
         u // temporary views take precedence over catalog table names
 
-      case u @ UnresolvedRelation(CatalogObjectIdentifier(Some(catalogPlugin), ident)) =>
-        loadTable(catalogPlugin, ident).map(DataSourceV2Relation.create).getOrElse(u)
+      case u @ UnresolvedRelation(CatalogObjectIdentifier(maybeCatalog, ident)) =>
+        // First try loading the table with a loadable catalog, then fallback to the session
+        // catalog if that exists
+        maybeCatalog.flatMap(loadTable(_, ident))
```

Comment:
How about?
```
        maybeCatalog.orElse(sessionCatalog)
          .flatMap(loadTable(_, ident))
```

Reply:
That's the tricky thing. We need to see if we can load using the defined catalog first. If the catalog is defined, but it doesn't return a table, then we miss out on using the session catalog.

Reply:
When a user tries to use table `cat1.ns1.ns2.tbl`, where
1. catalog `cat1` exists, but without table `ns1.ns2.tbl`
2. v2 session catalog is specified and it has a table named `cat1.ns1.ns2.tbl`

IMHO, Spark should throw no table found exception. This is consistent with the follow statement in #24768:

> A session catalog should be used when the v1 catalog is responsible for tables with no catalog in the table identifier.

 Maybe I missed some discussions about fallback?

Reply:
> That's the tricky thing. We need to see if we can load using the defined catalog first. If the catalog is defined, but it doesn't return a table, then we miss out on using the session catalog.

I disagree.

Determining which catalog is responsible for an identifier is independent of the catalog results. If `maybeCatalog` is `None`, then the session catalog is responsible for the identifier. Whether to use v1 or v2 after that point depends on the provider.

Reply:
@rdblue Imagine the following use case. Let's say as the data science team, I have a database on the Hive MetaStore called `prod`. I have tables named like: `prod.tbl1`, `prod.tbl2`. Now, some other team (maybe data engineering team), creates a v2 catalog called `prod` and adds it to my environment.

Won't the `prod` catalog hijack all my requests, and start failing to resolve the tables I had declared before, which I was expecting the V2SessionCatalog to resolve? I, as an unaware user, suddenly I have all my jobs failing and chaos ensues.

Reply:
Yes. A conflict could break queries in both directions when there is a conflict between namespace `prod` and catalog `prod`:
* If a namespace takes precedence and a new one is created, it will break queries that use the catalog
* If a catalog takes precedence and a new one is created, it will break queries that use the namespace

When we wrote the SPIP, the choice was to make catalog take precedence because:
1. Any user can create a namespace globally in the catalog, possibly breaking other users
2. Users can't globally create catalogs -- that's done by administrators -- so the impact is limited to their own jobs where they can add Spark configs
3. Global catalogs are created by administrators and this happens more rarely

In short, we expect fewer problems when catalogs take precedence.